File: day14/day_fourteen.cpp
# include <iostream>

using namespace std;


void allPair(string str,string curr,int index)
{
	if(str.length() == index)
	{
		cout << curr << " ";
		return;
	}
	allPair(str,curr,index+1);
	allPair(str,curr + str[index],index+1);
}

void permute(string str,int l,int r)
{
	if(l == r)
	{
		cout << str << " ";
	}
	else
	{
		for(int i=l;i<=r;i++)
		{
			swap(str[l],str[i]);
			permute(str,l+1,r);
			swap(str[l],str[i]);
		}
	}
}

int main()
{
	string str = "ABC";
	allPair(str,"",0);
	cout << endl;
	permute(str,0,str.length()-1);
	cout << endl;
	return 0;
}

File: day14/dayFourteen.java
public class dayFourteen
{
	public static void main(String args[])
	{
		String str = "ABC";
		dayFourteen stringMethods = new dayFourteen();
		stringMethods.allPair(str,"",0);
		System.out.println();
		stringMethods.permute(str,0,str.length()-1);
		System.out.println();
	}

	
	public void allPair(String str,String curr,int index)
	{
		if(str.length() == index)
		{
			System.out.print(curr + " ");
			return;
		}

		allPair(str,curr,index+1);
		allPair(str,curr+str.charAt(index),index+1);
	}

	public String swap(String str, int i, int j) 
    {
    	char[] charArr = str.toCharArray(); 
        char temp = charArr[i] ; 
        charArr[i] = charArr[j]; 
        charArr[j] = temp; 
        return new String(charArr); 
    }

	public void permute(String str,int l,int r)
	{
		if(l == r)
		{
			System.out.print(str + " ");
		}
		else
		{
			for(int i=l;i<=r;i++)
			{
				str = swap(str,l,i);
				permute(str,l+1,r);
				str = swap(str,l,i);
			}
		}
	}
}